packages/wallet-sdk/src/connection/ServerMessage.ts

```diff
@@ -3,87 +3,66 @@
 
 import { IntNumber } from '../types';
 
-export interface ServerMessage {
-  type: string;
-  id?: IntNumber;
-}
+export type ServerMessage<T extends Type = Type> = Extract<_ServerMessage, { type: T }>;
 
-export interface ServerMessageOK extends ServerMessage {
-  type: 'OK';
-  id: IntNumber;
-  sessionId: string;
-}
+export type ServerMessageType = Type;
+type Type = _ServerMessage['type'];
 
-export interface ServerMessageFail extends ServerMessage {
-  type: 'Fail';
-  id: IntNumber;
-  sessionId: string;
-  error: string;
-}
-
-export function isServerMessageFail(msg: any): msg is ServerMessageFail {
-  return (
-    msg &&
-    msg.type === 'Fail' &&
-    typeof msg.id === 'number' &&
-    typeof msg.sessionId === 'string' &&
-    typeof msg.error === 'string'
-  );
-}
-
-export interface ServerMessageIsLinkedOK extends ServerMessage {
-  type: 'IsLinkedOK';
-  id: IntNumber;
-  sessionId: string;
-  linked: boolean;
-  onlineGuests: number;
-}
-
-export interface ServerMessageLinked extends ServerMessage {
-  type: 'Linked';
-  sessionId: string;
-  onlineGuests: number;
-}
-
-export interface ServerMessageGetSessionConfigOK extends ServerMessage {
-  type: 'GetSessionConfigOK';
-  id: IntNumber;
-  sessionId: string;
-  webhookId: string;
-  webhookUrl: string;
-  metadata: { [field: string]: string };
-}
-
-export interface ServerMessageSessionConfigUpdated extends ServerMessage {
-  type: 'SessionConfigUpdated';
-  sessionId: string;
-  webhookId: string;
-  webhookUrl: string;
-  metadata: { [field: string]: string };
-}
-
-export interface ServerMessagePublishEventOK extends ServerMessage {
-  type: 'PublishEventOK';
-  id: IntNumber;
-  sessionId: string;
-  eventId: string;
-}
-
-export interface ServerMessageEvent extends ServerMessage {
-  type: 'Event';
-  sessionId: string;
-  eventId: string;
-  event: string;
-  data: string;
-}
-
-export function isServerMessageEvent(msg: any): msg is ServerMessageEvent {
-  return (
-    msg &&
-    msg.type === 'Event' &&
-    typeof msg.sessionId === 'string' &&
-    typeof msg.eventId === 'string' &&
-    typeof msg.event === 'string' &&
-    typeof msg.data === 'string'
-  );
-}
+type _ServerMessage =
+  | {
+      type: 'Heartbeat'; // TODO: remove. it's client side only virtual message.
+    }
+  | {
+      type: 'OK';
+      id: IntNumber;
+      sessionId: string;
+    }
+  | {
+      type: 'Fail';
+      id: IntNumber;
+      sessionId: string;
+      error: string;
+    }
+  | {
+      type: 'IsLinkedOK';
+      id: IntNumber;
+      sessionId: string;
+      linked: boolean;
+      onlineGuests: number;
+    }
+  | {
+      type: 'Linked';
+      id?: IntNumber;
```

Comment:
for some reason it was optional. keeping it that way...
![image](https://github.com/coinbase/coinbase-wallet-sdk/assets/966662/eb3bb0cf-e719-4864-8082-e76454103224)